news-bundle/src/Controller/Page/NewsFeedController.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\NewsBundle\Controller\Page;
+
+use Contao\CoreBundle\Asset\ContaoContext;
+use Contao\CoreBundle\Controller\AbstractController;
+use Contao\CoreBundle\Routing\Page\DynamicRouteInterface;
+use Contao\CoreBundle\Routing\Page\PageRoute;
+use Contao\CoreBundle\ServiceAnnotation\Page;
+use Contao\NewsBundle\Event\FetchArticlesForFeedEvent;
+use Contao\NewsBundle\Event\TransformArticleForFeedEvent;
+use Contao\PageModel;
+use FeedIo\Feed;
+use FeedIo\Specification;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * @Page(contentComposition=false)
+ */
+class NewsFeedController extends AbstractController implements DynamicRouteInterface
+{
+    final public const TYPE = 'news_feed';
+
+    public static $contentTypes = [
+        'atom' => 'application/atom+xml',
+        'json' => 'application/feed+json',
+        'rss' => 'application/rss+xml',
+    ];
+
+    private array $urlSuffixes = [
+        'atom' => '.xml',
+        'json' => '.json',
+        'rss' => '.xml',
+    ];
+
+    public function __construct(private readonly ContaoContext $contaoContext, private readonly Specification $specification)
+    {
+    }
+
+    public function __invoke(Request $request, PageModel $pageModel): Response
+    {
+        $this->initializeContaoFramework();
+
+        $staticUrl = $this->contaoContext->getStaticUrl();
+        $baseUrl = $staticUrl ?: $request->getSchemeAndHttpHost();
+
+        $feed = (new Feed())
+            ->setTitle($pageModel->title)
+            ->setDescription($pageModel->description)
+            ->setLanguage($pageModel->language)
+        ;
+
+        $event = new FetchArticlesForFeedEvent($feed, $request, $pageModel);
+        $this->container->get('event_dispatcher')->dispatch($event);
+
+        if (null === ($articles = $event->getArticles())) {
+            return new Response('', Response::HTTP_NO_CONTENT);
```

Comment:
is it correct to render an empty response? Or should we render an empty feed (if feedio supports that)? Also, this should be a 200 not 204 status code.

> https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204
>
> The HTTP 204 No Content success status response code indicates that a request has succeeded, but that the client doesn't need to navigate away from its current page.
> 
> This might be used, for example, when implementing "save and continue editing" functionality for a wiki site. > In this case a [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT) request would be used to save the page, and the 204 No Content response would be sent to indicate that the editor should not be replaced by some other page.The HTTP 204 No Content success status response code indicates that a request has succeeded, but that the client doesn't need to navigate away from its current page.

Reply:
Yes, I agree. The feed itself exists, it just does not have any articles. So returning an empty feed seems right to me. Fixed in 699713dfe